cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java

```diff
@@ -447,11 +466,13 @@ public SparkConfig generateSparkSubmitConfig(SparkSubmitContext sparkSubmitConte
     // Add spark pod labels. This will be same as job labels
     populateLabels(sparkConfMap);
 
+    CompletableFuture<Boolean> podStatusFuture = new CompletableFuture<>();
+    this.podWatcherCancellable = watchForPodStatus(podStatusFuture);
```

Comment:
It doesn't seem like this is the right place to start a watcher thread, since we're just creating a config here. It assumes that the KubeMasterEnvironment will not be destroyed before the spark job ends and that this method won't be called more than once. In general is a bit unexpected for a method like this to have side effects.

I think it would be better if the SparkConfig that is returned here has a method to get a Thread that watches the Spark job and have the MasterEnvironmentSparkSubmitter in charge of starting and stopping it.

Reply:
Or potentially to add something to MasterEnvironment to get a watcher thread

Reply:
Makes sense, let me move start/stop logic to `MasterEnvironmentSparkSubmitter` like you have suggested above.